Localize the Countries on the A4A signup form

d-alleyne opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to Automattic/i18n-issues#828

## Proposed Changes

Localize the country names on the signup form.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Improve the signup experience.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Run A4A dev environment locally with `yarn start-a8c-for-agencies`.
2. Open `/signup?locale=ja` (the locale param can be any Mag-16 language slug).
3. Verify that the countries list is in Japanese.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

d-alleyne committed:
Localize the Countries on the A4A signup form

d-alleyne committed:
Update the test

yuliyan commented on client/landing/stepper/hooks/use-countries.ts:
```diff
@@ -4,11 +4,15 @@ import wpcom from 'calypso/lib/wp';
-		queryFn: () => wpcom.req.get( '/woocommerce/countries/regions/', { apiNamespace: 'wpcom/v2' } ),
+		queryKey: [ 'countries', localeSlug ],
+		queryFn: () =>
+			wpcom.req.get( `/woocommerce/countries/regions/?_locale=${ localeSlug }`, {
```
On inspecting the network requests, I've noticed that with this change the `_locale` parameter is now being added twice. Here's the URL of the request that I'm seeing:

`https://public-api.wordpress.com/wpcom/v2/woocommerce/countries/regions/?_locale=es&_envelope=1&_locale=es`

yuliyan commented on client/landing/stepper/hooks/use-countries.ts:
```diff
@@ -4,11 +4,15 @@ import wpcom from 'calypso/lib/wp';
 type WooCountries = Record< string, string >;
 
 export function useCountries(
+	localeSlug: string | null = 'en',
```
I found that the the `useCountries` hook is used in a few other places outside of the `useCountriesAndStates` hook, which would cause them to default to `en` locale slug.

d-alleyne replied:
I believe this is OK, considering that it currently returns countries in English.

yuliyan replied:
Do we know for sure that we currently have the same behavior anywhere `useCountries()` is used? The `wpcom` client is [expected to set the `_locale` paramater](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/wp/localization/index.js#L22), but it seems that it fails to in this case due to a race condition. However, I can't tell for sure if that's a consistent behavior at the moment because I'm unfamiliar with the other use cases, but it seems very unlikely.

In any case, with the currently implementation, when a parameter is not provided, it results in appending `?_locale=en` which is at the very least redundant.

d-alleyne replied:
@yuliyan Is it possible that the `useCountries` hook is caching the first (English?) request, and thus we need a [queryKey](https://tanstack.com/query/latest/docs/framework/react/guides/query-keys#if-your-query-function-depends-on-a-variable-include-it-in-your-query-key) to ensure localized countries are cached correctly? Could you also elaborate on how the race condition occurs?

Additionally, using private mode in both Brave and Firefox on [production](https://agencies.automattic.com/signup?locale=ja) shows a single request to the endpoint, and it doesn't include the _locale parameter.

yuliyan replied:
> Could you also elaborate on how the race condition occurs?

The race condition I'm referring to is between the [setupLocale()](https://github.com/Automattic/wp-calypso/blob/trunk/client/boot/common.js#L427) and [wpcom.req.get()](https://github.com/Automattic/wp-calypso/blob/add/localize-a4a-countries-signup-form/client/landing/stepper/hooks/use-countries.ts#L13) request triggered by the `useCountries()` hook. The former is initiated first, but it appears that the locale setup is still not finished when the `wpcom.req.get()` is triggered which results in requesting the countries with the English locale slug.

Having that in mind and the fact that [`_locale` parameter is omitted for English](https://github.com/Automattic/wp-calypso/blob/add/localize-a4a-countries-signup-form/client/lib/wp/localization/index.js#L13), would explain why the request on production doesn't include the parameter.

> @yuliyan Is it possible that the useCountries hook is caching the first (English?) request, and thus we need a [queryKey](https://tanstack.com/query/latest/docs/framework/react/guides/query-keys#if-your-query-function-depends-on-a-variable-include-it-in-your-query-key) to ensure localized countries are cached correctly

It's cached and there's also persistent caching, so I think it makes sense to include the locale slug in the query key, as  you've done in your implementation.

yuliyan replied:
I looked for other instances that use the `useCountries()` hook and found the following flow - `wordpress.com/setup/plugin-bundle/storeAddress?siteSlug={SITE}`, where `{SITE}` is a newly created site with Creator plan and WooCommerce theme.

It appears to be working correctly on production, requesting the countries in the user locale. However, with the changes introduces in this PR, the `_locale` is included twice - `https://public-api.wordpress.com/wpcom/v2/woocommerce/countries/regions/?_locale=en&_envelope=1&_locale=ja` - the first one is added as a result of the default value `en`, whereas the second one is from the [WPCOM API client](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/wp/localization/index.js#L22) and has the correct user locale.

yuliyan commented on client/a8c-for-agencies/sections/signup/agency-details-form/hooks/use-countries-and-states.ts:
```diff
@@ -47,7 +47,8 @@ export interface Option {
 
 export function useCountriesAndStates() {
-	const { data: countriesList } = useCountries();
+	const localeSlug = getLocaleSlug();
```
Having `useTranslate()` hook below will ensure re-rendering when the locale data changes, and thus the static `getLocaleSlug()` should work as expected, but for the sake of good practices, I'd suggest using `useLocale()` or read the `localeSlug` property from `useTranslate()`.

d-alleyne committed:
Update the useCountries hook to be aware that the locale affects the results

Currently, the `useQuery` call within the `useCountries` hook is
unaware that the `wpcom.req.get` request is altered to append `_locale`
when the user's locale is not English.

yuliyan commented on client/landing/stepper/hooks/test/use-countries.jsx:
```diff
@@ -21,7 +21,7 @@ describe( 'use-countries hook', () => {
 
 		nock( 'https://public-api.wordpress.com' )
-			.get( '/wpcom/v2/woocommerce/countries/regions/' )
+			.get( '/wpcom/v2/woocommerce/countries/regions/?_locale=en' )
```
`?_locale=en` is no longer needed and is causing the tests to fail.

yuliyan commented on client/a8c-for-agencies/sections/signup/agency-details-form/hooks/use-countries-and-states.ts:
```diff
@@ -47,8 +47,8 @@ export interface Option {
 export function useCountriesAndStates() {
-	const { data: countriesList } = useCountries();
 	const translate = useTranslate();
+	const { data: countriesList } = useCountries();
```
Nit: This seems like unnecessary change, so it can be removed to reduce the noise.

yuliyan commented on client/landing/stepper/hooks/use-countries.ts:
```diff
@@ -6,9 +7,13 @@ type WooCountries = Record< string, string >;
-		queryFn: () => wpcom.req.get( '/woocommerce/countries/regions/', { apiNamespace: 'wpcom/v2' } ),
+		queryKey: [ 'countries', localeSlug ],
+		queryFn: () =>
+			wpcom.req.get( `/woocommerce/countries/regions/`, {
```
Nit: The use of template string seems unnecessary.

yuliyan reviewed (approved):
Thank you for making the changes! Tested and branch and it seems to be working as expected.

The PR should be good to merge after fixing the failing unit tests.

d-alleyne committed:
Reset test

d-alleyne committed:
Revert unneeded changes
